### PR TITLE
serializer: extended inventory csv export

### PR DESCRIPTION
**Pull request by rerowep**
* Adds an extented inventory csv export `csvext`.
* Adds a cli for this export.
* Needed by RT_000455

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- Which task/US does it implement?
- Which issue does it fix?

## Dependencies

My PR depends on the following `rero-ils-ui`'s PR(s):

* rero/rero-ils-ui#<xx>

## How to test?

- What command should I have to run to test your PR?
- What should I test through the UI?

### CODE REVIEW

File: rero_ils/modules/cli/utils.py [hunk 1]
```diff
@@ -1691,3 +1692,46 @@ def add_cover_urls(verbose):
...
+
+
+@utils.command()
+@click.argument('outfile', type=click.File('w'))
+@click.option('-o', '--organisation', multiple=True)
+@click.option('-l', '--library', multiple=True)
+@click.option('-L', '--location', multiple=True)
+@with_appcontext
+def export_csv(outfile, organisation, library, location):
+    """Delete loans with state CREATED."""
```
Comment (Garfield-fr) [inline rero_ils/modules/cli/utils.py:1704]: The docstring is not correct

File: rero_ils/modules/cli/utils.py [hunk 2]
```diff
@@ -1691,3 +1692,46 @@ def add_cover_urls(verbose):
 def delete_loans_created(verbose, hours):
     """Delete loans with state CREATED."""
     return task_delete_loans_created(verbose=verbose, hours=hours)
+
+
+@utils.command()
+@click.argument('outfile', type=click.File('w'))
```
Comment (Garfield-fr) [inline rero_ils/modules/cli/utils.py:1698]: `file` ?

File: rero_ils/modules/items/serializers/collector_ext.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+                if authorized_access_point in contribution.get('agent'):
+                    creator.append(
+                        f'{contribution["agent"][authorized_access_point]}'
+                        f', {contribution.get("role")}'
+                    )
+            document_data['document_contribution'] = ' ; '.join(creator)
+            document_main_type = []
+            document_sub_type = []
+            for document_type in data.get('type'):
+                # data = document_type.to_dict()
```
Comment (Garfield-fr) [inline rero_ils/modules/items/serializers/collector_ext.py:87]: remove this

File: rero_ils/modules/items/serializers/collector_ext.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+            return document_data
+
+        doc_search = DocumentsSearch() \
+            .filter('terms', holdings__items__pid=list(item_pids))
+            # .source(['pid', 'title', 'contribution', 'provisionActivity',
+            #          'type', '_masked', 'isbn', 'issn', 'identifiedBy',
+            #          'seriesStatement', 'editionStatement',
+            #          'responsibilityStatement', 'extent', 'tableOfContents'
+            #          'note', 'credits', 'work_access_point',
+            #          'contentMediaCarrier'])
```
Comment (Garfield-fr) [inline rero_ils/modules/items/serializers/collector_ext.py:206]: Remove comments

**Review by Garfield-fr — APPROVED** [review]

File: rero_ils/modules/items/serializers/collector_ext.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+        :param hit: item ES record.
+        :param csv_data: dictionary of data.
+        :param loans: loans data.
+        """
+        stat = next(items_stats)
+        csv_data['item_checkouts_count'] = stat.get('checkout_count', 0)
+        csv_data['item_renewals_count'] = stat.get('renewal_count', 0)
+        csv_data['last_transaction_date'] = stat.get('last_transaction')
+        csv_data['last_checkout_date'] = stat.get('last_checkout')
```
Comment (lauren-d) [inline rero_ils/modules/items/serializers/collector_ext.py:426]: It appears that only `get_documents_by_item_pids()` is overloaded.
You could remove all these lines of code if you inherit of Collector class

File: rero_ils/modules/items/serializers/csv_ext.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+from rero_ils.modules.serializers import CachedDataSerializerMixin
+from rero_ils.utils import get_i18n_supported_languages
+
+from .collector_ext import CollecterExt as Collecter
+
+
+class ItemExtCSVSerializer(CSVSerializer, CachedDataSerializerMixin):
+    """Serialize item search for csv."""
+
+    def serialize_search(self, pid_fetcher, search_result, links=None,
```
Comment (lauren-d) [inline rero_ils/modules/items/serializers/csv_ext.py:39]: I have a suggestion regarding this class. but not sure it works.

First of all, add an attribute `collector` to  `ItemCSVSerializer` class
```python

from .collector import Collecter

class ItemCSVSerializer(CSVSerializer, CachedDataSerializerMixin):
    """Serialize item search for csv."""

    collector = Collecter
```
Then, modify code in `ItemCSVSerializer.ItemCSVSerializer()` to use new attribute
``` python
...
for pids, batch_results in self.collector.batch(results=search_result):
  # get documents
  documents = self.collector.get_documents_by_item_pids(
      item_pids=pids, language=language)
  # get loans
  items_stats = self.collector.get_loans_by_item_pids(item_pids=pids)
  for hit in batch_results:
      csv_data = self.collector.get_item_data(hit)
      # _process_item_types_libs_locs(self, csv_data)
      _process_item_types_libs_locs(csv_data)
      self.collector.append_document_data(csv_data, documents)
      self.collector.append_local_fields(csv_data)
      # update csv data with loan
      self.collector.append_loan_data(hit, csv_data, items_stats)
      self.collector.append_issue_data(hit, csv_data)
      # write csv data
      data = self.process_dict(dictionary=csv_data)
      writer.writerow(data)
      yield line.read()
...
```
Finally, you can do this in `ItemExtCSVSerializer` class:
```python
...
from .collector_ext import CollecterExt

from .collector import Collecter

class ItemExtCSVSerializer(ItemCSVSerializer, CachedDataSerializerMixin):
    """Serialize item search for csv."""

    collector = CollecterExt
```

**Review by lauren-d — CHANGES_REQUESTED** [review]

File: rero_ils/modules/cli/utils.py (hunk 2 above)
Comment (rerowep) [reply rero_ils/modules/cli/utils.py:1698]: I allways make a difference between `infile`, `outile` etc. This name is seen by `--help`.

**Review by lauren-d — APPROVED** [review]

**Review by benerken — APPROVED** [review]

File: rero_ils/modules/items/serializers/__init__.py [hunk 7]
```diff
@@ -100,13 +101,105 @@
...
 """CSV serializer."""
 csv_item_response = record_responsify(_csv, "text/csv")
 csv_item_search = search_responsify_file(
     _csv, 'text/csv',
     file_extension='csv',
     file_suffix='inventory'
 )
+csv_item_ext_response = record_responsify(_csv_ext, "text/csv+ext")
+csv_item_ext_search = search_responsify_file(
+    _csv_ext, 'text/csv+ext',
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/__init__.py:199]: remarks: What's the limit if this approach ? If other customers asked for specific report, we could imagine `text/csv+customerX_outputY` ?

File: rero_ils/modules/items/serializers/collector_ext.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+from rero_ils.modules.documents.api import DocumentsSearch
+from rero_ils.modules.documents.utils import title_format_text_head
+from rero_ils.modules.documents.views import work_access_point
+
+from .collector import Collector
+
+
+class CollectorExt(Collector):
+    """collect data for csv."""
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector_ext.py:28]: Could be interesting to add more docstring to explain the usage and purpose of the class. What's the difference with classic Item CSV export

File: rero_ils/modules/items/serializers/collector_ext.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            for identified_by in data.get('identifiedBy', []):
+                identified_type = identified_by.get("type")
+                if identified_type not in ['bf:Isbn', 'bf:Issn']:
+                    identified = (
+                        f'({identified_type}){identified_by.get("value")}')
+                    for additional in [
+                            "source", "status", "qualifier", "note"]:
+                        if more := identified_by.get(additional):
+                            identified = f'{identified}; {additional}: {more}'
+                    identifiers.append(identified)
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector_ext.py:90]: We create a "framework" to deal with document identifier to skip all these scripted lines.
```python
identifiers = [
    IdentifierFactory\
        .create_identifier(data)\
        .render(render_class=MyCustomIdentifierRendered())
    for data in self.get('identifiedBy', [])
    if data.get('type') in [IdentifierType.ISBN, IdentifierType.ISSN]
]
```
See `rero_ils/commons/identifiers.py`

File: rero_ils/modules/items/serializers/collector_ext.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                cls.separator.join(content_media_carriers)
+
+            return document_data
+
+        doc_search = DocumentsSearch() \
+            .filter('terms', holdings__items__pid=list(item_pids))
+        docs = {}
+        for doc in doc_search.scan():
+            docs[doc.pid] = _build_doc(doc.to_dict())
+        return docs
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector_ext.py:182]: Could be reduces 
```python
return { hit.pid: _build_doc(hit.to_dict()) for hit in search.scan() }
```

File: rero_ils/modules/items/serializers/collector_ext.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+                            cmc_data.append(', '.join(value))
+                        else:
+                            cmc_data.append(value)
+                content_media_carriers.append('; '.join(cmc_data))
+            document_data['document_contentMediaCarrier'] = \
+                cls.separator.join(content_media_carriers)
+
+            return document_data
+
+        doc_search = DocumentsSearch() \
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector_ext.py:177]: maybe use only `search` as name

File: rero_ils/modules/items/serializers/csv.py [hunk 12]
```diff
@@ -30,12 +30,14 @@
...
 from rero_ils.utils import get_i18n_supported_languages
 
-from .collector import Collecter
+from .collector import Collector
 
 
 class ItemCSVSerializer(CSVSerializer, CachedDataSerializerMixin):
     """Serialize item search for csv."""
 
+    collector = Collector
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/csv.py:39]: Why ? What's the purpose of store the `Collector` class reference as `self.collector`. Simply use `Collector.my_requested_method()`

File: rero_ils/modules/items/serializers/csv.py [hunk 13]
```diff
@@ -86,21 +88,23 @@ def _process_item_types_libs_locs(csv_data):
             writer.writeheader()
             yield line.read()
 
-            for pids, batch_results in Collecter.batch(results=search_result):
+            for pids, batch_results in self.collector.batch(
+                results=search_result):
                 # get documents
-                documents = Collecter.get_documents_by_item_pids(
+                documents = self.collector.get_documents_by_item_pids(
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/csv.py:94]: I don't understand this change (and next)

**Review by zannkukai — CHANGES_REQUESTED** [review]
Missing unitest

**Pull request closed without merging**